core/EE_Registry.core.php

```diff
@@ -164,53 +176,81 @@ class EE_Registry implements ResettableInterface
      */
     protected $_cache_on = true;
 
+    /**
+     * @var ObjectIdentifier
+     */
+    private $object_identifier;
 
 
     /**
      * @singleton method used to instantiate class object
-     * @param  EE_Dependency_Map $dependency_map
+     * @param EE_Dependency_Map|null   $dependency_map
+     * @param Mirror|null              $mirror
+     * @param ClassInterfaceCache|null $class_cache
+     * @param ObjectIdentifier|null    $object_identifier
      * @return EE_Registry instance
-     * @throws InvalidArgumentException
-     * @throws InvalidInterfaceException
-     * @throws InvalidDataTypeException
      */
-    public static function instance(EE_Dependency_Map $dependency_map = null)
-    {
+    public static function instance(
+        EE_Dependency_Map $dependency_map = null,
+        Mirror $mirror = null,
+        ClassInterfaceCache $class_cache = null,
+        ObjectIdentifier $object_identifier = null
+    ) {
         // check if class object is instantiated
-        if (! self::$_instance instanceof EE_Registry) {
-            self::$_instance = new self($dependency_map);
+        if (
+            ! self::$_instance instanceof EE_Registry
+            && $dependency_map instanceof EE_Dependency_Map
+            && $mirror instanceof Mirror
+            && $class_cache instanceof ClassInterfaceCache
+            && $object_identifier instanceof ObjectIdentifier
+        ) {
+            self::$_instance = new self(
+                $dependency_map,
+                $mirror,
+                $class_cache,
+                $object_identifier
+            );
         }
         return self::$_instance;
     }
 
 
-
     /**
      * protected constructor to prevent direct creation
      *
      * @Constructor
-     * @param  EE_Dependency_Map $dependency_map
-     * @throws InvalidDataTypeException
-     * @throws InvalidInterfaceException
-     * @throws InvalidArgumentException
-     */
-    protected function __construct(EE_Dependency_Map $dependency_map)
-    {
-        $this->_dependency_map = $dependency_map;
+     * @param  EE_Dependency_Map  $dependency_map
+     * @param Mirror              $mirror
+     * @param ClassInterfaceCache $class_cache
+     * @param ObjectIdentifier    $object_identifier
+     */
+    protected function __construct(
+        EE_Dependency_Map $dependency_map,
+        Mirror $mirror,
+        ClassInterfaceCache $class_cache,
+        ObjectIdentifier $object_identifier
+    ) {
```

Comment:
Is EE_Registry _only_ instantiated in core?  If so, this is fine as is.  But if not, I wonder if the new args should default to null and we use LoaderFactory to instantiate if they are null?  I prefer to _not_ have to do that if possible but just pointing it out in case.

Reply:
only in core.
ideally, your DI container (which is kinda sorta what EE_Registry is behaving like) is constructed before anything else and is the source for all other objects.